Implement browser.test.addTest to support web platform testing of extensions

elijahsawyers opened the pull request:
#### 5ca13e0fb7c00fa541cfcfe33fbe3024bdb45212
<pre>
Implement browser.test.addTest to support web platform testing of extensions
<a href="https://bugs.webkit.org/show_bug.cgi?id=287566">https://bugs.webkit.org/show_bug.cgi?id=287566</a>
<a href="https://rdar.apple.com/144706714">rdar://144706714</a>

Reviewed by Timothy Hatcher.

This patch adds a new API for running tests that are represented by promises.
If any browser.test assertion within the test method fails, we reject the
promise and notify fail; otherwise, we resolve the promise and notify pass.
It&apos;s worth noting that tests are ran in the order in which they&apos;re added with
the addTest method.

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
Define the delegate methods that are called when the following messages are received...

* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.messages.in:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionControllerAPITestCocoa.mm:
(WebKit::WebExtensionController::testAdded):
(WebKit::WebExtensionController::testStarted):
...the two new messages are:

the testAdded message, which is sent when browser.test.addTest is called,
and the testStarted message, which is called when the next test is removed
from the queue and has started executing.

* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITest.idl:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
(WebKit::WebExtensionAPITest::assertTrue):
(WebKit::WebExtensionAPITest::assertDeepEq):
(WebKit::WebExtensionAPITest::assertEquals):
(WebKit::WebExtensionAPITest::assertEq):
(WebKit::WebExtensionAPITest::addTest):
(WebKit::WebExtensionAPITest::startNextTest):
(WebKit::assertEquals): Deleted.
Implement the logic for browser.test.addTest, which just maintains a queue of
test methods to run and resolves/rejects the promise based on whether or not
any assertions are hit during the execution.

* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITest.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPITest, AddTestThatPasses)):
(TestWebKitAPI::TEST(WKWebExtensionAPITest, AddTestThatFails)):
Add new tests to ensure that passing and failing tests resolve and reject,
respectively. Additionally, check that the new delegate methods are called
as expected.

* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager initForExtension:extensionControllerConfiguration:]):
(-[TestWebExtensionManager _webExtensionController:recordTestAssertionResult:withMessage:andSourceURL:lineNumber:]):
(-[TestWebExtensionManager _webExtensionController:recordTestEqualityResult:expectedValue:actualValue:withMessage:andSourceURL:lineNumber:]):
(-[TestWebExtensionManager _webExtensionController:recordTestAddedWithTestName:andSourceURL:lineNumber:]):
(-[TestWebExtensionManager _webExtensionController:recordTestStartedWithTestName:andSourceURL:lineNumber:]):
(-[TestWebExtensionManager _webExtensionController:recordTestFinishedWithResult:message:andSourceURL:lineNumber:]):
Throw away test events if we&apos;re currently executing a test that was added
with the new browser.test.addTest API. We don&apos;t care what the result of
those assertions are, just that the promise resolves/rejects as expected.

Canonical link: <a href="https://commits.webkit.org/290558@main">https://commits.webkit.org/290558@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/51a06eace7398386eccbcb468d075b63d2b7b3c8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90398 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9928 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45325 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95404 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41174 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92450 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10315 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18246 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69598 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27164 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93399 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7899 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82003 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49944 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7633 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36364 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40306 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77968 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37437 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97236 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17586 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78588 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17843 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77827 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77800 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22250 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20859 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10857 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14226 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17596 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17337 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20789 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19121 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
```diff
@@ -57,6 +57,18 @@ WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA))
+- (void)_webExtensionController:(WKWebExtensionController *)controller recordTestStartedWithTestName:(NSString *)testName andSourceURL:(NSString *)sourceURL lineNumber:(unsigned)lineNumber;
+
 /*!
  @abstract Delegate for the `browser.test.notifyPass()` and `browser.test.notifyFail()` JavaScript testing APIs.
```
Add a note that this is called for `addTest` tests now too?

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -172,6 +172,9 @@
         return;
 
+    if (m_runningTest && !m_hitAssertion)
+        m_hitAssertion = !actualValue;
```
A simple helper for this might still be good. It can even be inline in the header.

```cpp
void recordAssertionIfNeeded(bool result)
{
    if (m_runningTest && !m_hitAssertion && !result)
        m_hitAssertion = true;
}
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +373,74 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
+
+    RefPtr page = toWebPage(context);
+    if (!page)
+        return [JSValue valueWithNewPromiseRejectedWithReason:toErrorString(nullString(), nullString(), "Error creating a new test."_s) inContext:toJSContext(context)];
```
```suggestion
        return [JSValue valueWithNewPromiseRejectedWithReason:toErrorString("test.addTest()"_s, nullString(), "Error creating a new test."_s) inContext:toJSContext(context)];
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +373,74 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
+
+    RefPtr webExtensionControllerProxy = page->webExtensionControllerProxy();
+    if (!webExtensionControllerProxy)
+        return [JSValue valueWithNewPromiseRejectedWithReason:toErrorString(nullString(), nullString(), "Error creating a new test."_s) inContext:toJSContext(context)];
```
```suggestion
        return [JSValue valueWithNewPromiseRejectedWithReason:toErrorString("test.addTest()"_s, nullString(), "Error creating a new test."_s) inContext:toJSContext(context)];
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +373,74 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
+    auto location = scriptLocation(context);
+    auto webExtensionControllerIdentifier = webExtensionControllerProxy->identifier();
+
+    m_testQueue.append(Test {
```
Shouldn't need the type.
```suggestion
    m_testQueue.append({
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +373,74 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
     return assertResolves(context, result, message);
 }
 
+JSValue *WebExtensionAPITest::addTest(JSContextRef context, JSValue *testFunction, NSString *testName)
```
You can infer the test name from the function, which would be a good way to add tests with just one argument. 

```cpp
if (!testName.length)
    testName = testFunction[@"name"].toString;
```

`testName` would also be better as the first argument and optional.

```javascript
browser.test.addTest('runtime.getURL', async () => { })
```

So the function can flow inline after the test name (if not provided by the function name).

And this is still valid:

```javascript
browser.test.addTest(async runtimeGetURL() => { })
```

elijahsawyers replied:
We chatted offline and decided that we're only going to allow specifying the test name with the function name. If the function is anonymous, we'll throw an error

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +373,74 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
+void WebExtensionAPITest::startNextTest()
+{
+    m_runningTest = true;
+    Test test = m_testQueue.first();
```
I think you can just take the test here.
```suggestion
    m_runningTest = true;

    auto test = m_testQueue.takeFirst();
```

elijahsawyers replied:
I originally had it this way so that if we added a `browser.test.remainingTests` property, it would have the correct value even while the last test was running. However, we're not taking that approach anymore, so I agree this is better.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +373,74 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
+
+    [testPromise _awaitThenableResolutionWithCompletionHandler:^(JSValue *result, JSValue *error) {
+        m_runningTest = false;
+        m_testQueue.removeFirst();
```
```suggestion
```

xeenon commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITest.idl:
```diff
@@ -82,4 +82,7 @@
     [NeedsScriptContext] any assertSafeResolve(any function, [Optional] DOMString message);
 
+    // Adds a test represented by a promise. If the promise is resolved, the test passes; if the promise is rejected, the test fails.
+    [NeedsScriptContext] any addTest(any function, DOMString testName);
```
Since we can infer the test name from the function, and this order allows inline functions to flow better.
```suggestion
    [NeedsScriptContext] any addTest([Optional] DOMString testName, any function);
```

xeenon commented on Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
```diff
@@ -355,8 +358,24 @@ - (void)_webExtensionController:(WKWebExtensionController *)controller receivedT
+    _testsStarted = [_testsStarted arrayByAddingObject:testName];
+}
+
 - (void)_webExtensionController:(WKWebExtensionController *)controller recordTestFinishedWithResult:(BOOL)result message:(NSString *)message andSourceURL:(NSString *)sourceURL lineNumber:(unsigned)lineNumber
```
This method likely should get a `testName` param, or have a new `testName` method in addition to this one. That way this can record the finish result per test by name too.

I think a dictionary like `testResults` would be good to update here when you have the test name.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,42 @@
+
+    [manager run];
+    EXPECT_NS_EQUAL(manager.get().testsAdded, @[ testName ]);
+    EXPECT_NS_EQUAL(manager.get().testsStarted, @[ testName ]);
```
```suggestion
    [manager run];

    EXPECT_NS_EQUAL(manager.get().testsAdded, @[ testName ]);
    EXPECT_NS_EQUAL(manager.get().testsStarted, @[ testName ]);
```

This is a little odd, since the tests have ended by the time we get here. But it does make sure the harness is working, so this is fine.

With my other suggestion, this turns into this:

```suggestion
    [manager run];

    EXPECT_NS_EQUAL(manager.get().testsAdded, @[ testName ]);
    EXPECT_NS_EQUAL(manager.get().testsStarted, @[ testName ]);
    EXPECT_NS_EQUAL(manager.get().testResults, @{ testName: @YES });
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,42 @@
+
+    [manager run];
+    EXPECT_NS_EQUAL(manager.get().testsAdded, @[ testName ]);
+    EXPECT_NS_EQUAL(manager.get().testsStarted, @[ testName ]);
```
Ditto from the other test.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,42 @@
+        @"  browser.test.assertTrue(false)",
+        [NSString stringWithFormat:@"}, '%@'))", testName],
+        @"  .then(() => browser.test.notifyPass())",
+        @"  .catch(() => browser.test.notifyFail())"
```
Wouldn't this pass on `catch` and fail on `then` since the assert is `false`? `notifyFail` should have a message on why.

elijahsawyers replied:
`then` and `catch` are added to the promise returned by `assertRejects`, so it's correct as I have it.

If the promise returned by `addTest` resolved, we'd hit the `catch` block and fail the test.

elijahsawyers replied:
Though we can get rid of the `assertRejects` call and notify pass/fail as you mentioned. I just felt like calling `assertRejects` had a clearer intent.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,42 @@
+        @"  browser.test.assertTrue(true)",
+        [NSString stringWithFormat:@"}, '%@'))", testName],
+        @"  .then(() => browser.test.notifyPass())",
+        @"  .catch(() => browser.test.notifyFail())"
```
`notifyFail` should have a message on why.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
Some other things to consider testing:
* Multiple calls to `addTest`, making sure they all run.
* Multiple tests, one passes and another fails. Do they still run if a previous one fails?
* Exception in a test function.
* A non-async test function.
* An exception in a non-async test function.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +373,74 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
+
+    WebProcess::singleton().send(Messages::WebExtensionController::TestStarted(test.testName, test.location.first, test.location.second), test.webExtensionControllerIdentifier);
+
+    JSValue *testPromise = [test.testFunction.get() callWithArguments:nil];
```
We should check for an exception after the call, see `assertThrows`.
```suggestion
    JSValue *testPromise = [test.testFunction.get() callWithArguments:nil];
    if (testFunction.exception) {
        // FIXME: Record fail and return rejected promise.
    }

```

elijahsawyers commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +370,85 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
+    JSValue *resolveCallback = test.resolveCallback.get();
+    JSValue *rejectCallback = test.rejectCallback.get();
+
+    auto testComplete = [this, protectedThis = Ref { *this }, test, resolveCallback, rejectCallback](JSValue *result, JSValue *error) {
```
@xeenon Is this capture list okay?

xeenon replied:
See my suggestion, needs `RetainPtr` for the ObjC callbacks.

elijahsawyers commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +370,85 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
+            testComplete(result, error);
+        }];
+    } else
+        testComplete(testPromise, test.testFunction.get().context.exception);
```
@xeenon Would this be better named as `testPromiseOrResult`? In the case of non-asynchronous functions, it's not a promise, it's just the result of calling the function.

xeenon replied:
Or just `result`.

elijahsawyers commented on Source/WebKit/UIProcess/Extensions/WebExtensionController.messages.in:
```diff
@@ -41,7 +41,9 @@ messages -> WebExtensionController {
-    [Validator=inTestingMode] TestFinished(bool result, String message, String sourceURL, unsigned lineNumber)
+    [Validator=inTestingMode] TestAdded(String testName, String sourceURL, unsigned lineNumber)
+    [Validator=inTestingMode] TestStarted(String testName, String sourceURL, unsigned lineNumber)
+    [Validator=inTestingMode] TestFinished(String testName, bool result, String message, String sourceURL, unsigned lineNumber)
```
Is there a better way to handle the nullable string here for the test name? We won't always have it. E.g. if `notifyPass` or `notifyFail` is called outside of a test being added with `addTest`.

xeenon replied:
Nope, `String` handles null internally. It can be checked with `isNull()` if needed.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
```diff
@@ -57,11 +57,24 @@ WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA))
+ @abstract Delegate for the `browser.test.addTest()` JavaScript testing API.
+ @discussion Default implementation logs a message to the system console that a test was started with `testName`. Test harnesses should use this to perform actions at the start of a test.
+ */
+- (void)_webExtensionController:(WKWebExtensionController *)controller recordTestStartedWithTestName:(NSString *)testName andSourceURL:(NSString *)sourceURL lineNumber:(unsigned)lineNumber;
```
I think this is still clear and not as redundant.
```suggestion
- (void)_webExtensionController:(WKWebExtensionController *)controller recordTestStartedWithName:(NSString *)testName andSourceURL:(NSString *)sourceURL lineNumber:(unsigned)lineNumber;
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
```diff
@@ -57,11 +57,24 @@ WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA))
+ @note This is also called with the test results of a test that was added with `browser.test.addTest()`.
  */
-- (void)_webExtensionController:(WKWebExtensionController *)controller recordTestFinishedWithResult:(BOOL)result message:(NSString *)message andSourceURL:(NSString *)sourceURL lineNumber:(unsigned)lineNumber;
+- (void)_webExtensionController:(WKWebExtensionController *)controller recordTestFinishedWithTestName:(NSString *)testName result:(BOOL)result message:(NSString *)message andSourceURL:(NSString *)sourceURL lineNumber:(unsigned)lineNumber;
```
Ditto.
```suggestion
- (void)_webExtensionController:(WKWebExtensionController *)controller recordTestFinishedWithName:(NSString *)testName result:(BOOL)result message:(NSString *)message andSourceURL:(NSString *)sourceURL lineNumber:(unsigned)lineNumber;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionControllerAPITestCocoa.mm:
```diff
@@ -102,23 +102,48 @@
+{
+    auto delegate = this->delegate();
+    if ([delegate respondsToSelector:@selector(_webExtensionController:recordTestStartedWithTestName:andSourceURL:lineNumber:)]) {
+        [delegate _webExtensionController:wrapper() recordTestStartedWithTestName:testName andSourceURL:sourceURL lineNumber:lineNumber];
```
```suggestion
    if ([delegate respondsToSelector:@selector(_webExtensionController:recordTestStartedWithName:andSourceURL:lineNumber:)]) {
        [delegate _webExtensionController:wrapper() recordTestStartedWithName:testName andSourceURL:sourceURL lineNumber:lineNumber];
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionControllerAPITestCocoa.mm:
```diff
@@ -102,23 +102,48 @@
+{
+    auto delegate = this->delegate();
+    if ([delegate respondsToSelector:@selector(_webExtensionController:recordTestFinishedWithTestName:result:message:andSourceURL:lineNumber:)]) {
+        [delegate _webExtensionController:wrapper() recordTestFinishedWithTestName:testName result:result message:message andSourceURL:sourceURL lineNumber:lineNumber];
```
```suggestion
    if ([delegate respondsToSelector:@selector(_webExtensionController:recordTestFinishedWithName:result:message:andSourceURL:lineNumber:)]) {
        [delegate _webExtensionController:wrapper() recordTestFinishedWithName:testName result:result message:message andSourceURL:sourceURL lineNumber:lineNumber];
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -236,6 +242,9 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
         actualJSONValue = [actualJSONValue stringByAppendingString:@" (different)"];
 
+    if (m_runningTest && !m_hitAssertion)
+        m_hitAssertion = !strictEqual;
```
```suggestion
    recordAssertionIfNeeded(!strictEqual);
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +370,85 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
+
+    auto testName = testFunction[@"name"].toString;
+    if (!testName.length)
+        return [JSValue valueWithNewPromiseRejectedWithReason:toErrorString("test.addTest()"_s, nullString(), "The supplied test method must be named."_s) inContext:toJSContext(context)];
```
```suggestion
        return [JSValue valueWithNewPromiseRejectedWithReason:toErrorString("test.addTest()"_s, nullString(), "The supplied test function must be named."_s) inContext:toJSContext(context)];
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +370,85 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
+    JSValue *resultPromise = [JSValue valueWithNewPromiseInContext:testFunction.context fromExecutor:^(JSValue *resolve, JSValue *reject) {
+        resolveCallback = resolve;
+        rejectCallback = reject;
+    }];
```
This should move below the early returns since it isn't used for those.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +370,85 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
+    JSValue *resolveCallback = test.resolveCallback.get();
+    JSValue *rejectCallback = test.rejectCallback.get();
+
+    auto testComplete = [this, protectedThis = Ref { *this }, test, resolveCallback, rejectCallback](JSValue *result, JSValue *error) {
```
```suggestion
    auto testComplete = [this, protectedThis = Ref { *this }, test, resolveCallback = RetainPtr { resolveCallback }, rejectCallback = RetainPtr { rejectCallback }](JSValue *result, JSValue *error) {
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +370,85 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
+            startNextTest();
+    };
+
+    if (test.testFunction.get()._isThenable) {
```
This should be:
```suggestion
    if (testPromise._isThenable) {
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -361,6 +370,85 @@ static void assertEquals(JSContextRef context, bool result, NSString *expectedSt
+
+    WebProcess::singleton().send(Messages::WebExtensionController::TestStarted(test.testName, test.location.first, test.location.second), test.webExtensionControllerIdentifier);
+
+    JSValue *testPromise = [test.testFunction.get() callWithArguments:nil];
```
No need to use `get()` for a call. Only dot-syntax.
```suggestion
    JSValue *testPromise = [test.testFunction callWithArguments:nil];
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,252 @@
+{
+    NSString *testName = @"passingTest";
+    auto *backgroundScript = Util::constructScript(@[
+        @"browser.test.assertResolves(browser.test.addTest(async function passingTest() {",
```
I just noticed that `assertResolves` and `assertSafe` incorrectly call `notifyFail` when an error is hit. Those should be `fail()` instead. Can you fix that?

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,252 @@
+
+TEST(WKWebExtensionAPITest, AddAsyncTestThatPasses)
+{
+    NSString *testName = @"passingTest";
```
```suggestion
    auto *testName = @"passingTest";
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,252 @@
+
+TEST(WKWebExtensionAPITest, AddAsyncTestThatFails)
+{
+    NSString *testName = @"failingTest";
```
```suggestion
    auto *testName = @"failingTest";
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,252 @@
+
+TEST(WKWebExtensionAPITest, AddAsyncTestThatThrows)
+{
+    NSString *testName = @"failingTest";
```
```suggestion
    auto *testName = @"failingTest";
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,252 @@
+
+TEST(WKWebExtensionAPITest, AddMultipleAsyncTestsThatPass)
+{
+    NSArray<NSString *> *testNames = @[ @"testA", @"testB" ];
```
```suggestion
    auto *testNames = @[ @"testA", @"testB" ];
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,252 @@
+
+TEST(WKWebExtensionAPITest, AddMultipleAsyncTestsWithFailure)
+{
+    NSArray<NSString *> *testNames = @[ @"testA", @"testB" ];
```
```suggestion
    auto *testNames = @[ @"testA", @"testB" ];
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,252 @@
+
+TEST(WKWebExtensionAPITest, AddTestThatPasses)
+{
+    NSString *testName = @"passingTest";
```
```suggestion
    auto *testName = @"passingTest";
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,252 @@
+
+TEST(WKWebExtensionAPITest, AddTestThatFails)
+{
+    NSString *testName = @"failingTest";
```
```suggestion
    auto *testName = @"failingTest";
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,252 @@
+
+TEST(WKWebExtensionAPITest, AddTestThatThrows)
+{
+    NSString *testName = @"failingTest";
```
```suggestion
    auto *testName = @"failingTest";
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,252 @@
+
+TEST(WKWebExtensionAPITest, AddMultipleTestsThatPass)
+{
+    NSArray<NSString *> *testNames = @[ @"testA", @"testB" ];
```
```suggestion
    auto *testNames = @[ @"testA", @"testB" ];
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,252 @@
+
+TEST(WKWebExtensionAPITest, AddMultipleTestsWithFailure)
+{
+    NSArray<NSString *> *testNames = @[ @"testA", @"testB" ];
```
```suggestion
    auto *testNames = @[ @"testA", @"testB" ];
```

xeenon commented on Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
```diff
@@ -355,8 +359,29 @@ - (void)_webExtensionController:(WKWebExtensionController *)controller receivedT
+    _testsAdded = [_testsAdded arrayByAddingObject:testName];
+}
+
+- (void)_webExtensionController:(WKWebExtensionController *)controller recordTestStartedWithTestName:(NSString *)testName andSourceURL:(NSString *)sourceURL lineNumber:(unsigned)lineNumber
```
```suggestion
- (void)_webExtensionController:(WKWebExtensionController *)controller recordTestStartedWithName:(NSString *)testName andSourceURL:(NSString *)sourceURL lineNumber:(unsigned)lineNumber
```

xeenon commented on Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
```diff
@@ -355,8 +359,29 @@ - (void)_webExtensionController:(WKWebExtensionController *)controller receivedT
+    _testsStarted = [_testsStarted arrayByAddingObject:testName];
+}
+
+- (void)_webExtensionController:(WKWebExtensionController *)controller recordTestFinishedWithTestName:(NSString *)testName result:(BOOL)result message:(NSString *)message andSourceURL:(NSString *)sourceURL lineNumber:(unsigned)lineNumber
```
```suggestion
- (void)_webExtensionController:(WKWebExtensionController *)controller recordTestFinishedWithName:(NSString *)testName result:(BOOL)result message:(NSString *)message andSourceURL:(NSString *)sourceURL lineNumber:(unsigned)lineNumber
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
```diff
@@ -218,6 +218,252 @@
+    auto manager = Util::loadExtension(manifest, @{ @"background.js": backgroundScript });
+
+    [manager run];
+    EXPECT_NS_EQUAL(manager.get().testsAdded, @[ ]);
```
I'd add an empty line between the run and `EXPECT_NS_EQUAL` for all these new tests.
```suggestion
    [manager run];

    EXPECT_NS_EQUAL(manager.get().testsAdded, @[ ]);
```

elijahsawyers committed:
Implement browser.test.addTest to support web platform testing of extensions
https://bugs.webkit.org/show_bug.cgi?id=287566
rdar://144706714

Reviewed by Timothy Hatcher.

This patch adds a new API for running tests that are represented by promises.
If any browser.test assertion within the test method fails, we reject the
promise and notify fail; otherwise, we resolve the promise and notify pass.
It's worth noting that tests are ran in the order in which they're added with
the addTest method.

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
Define the delegate methods that are called when the following messages are received...

* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.messages.in:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionControllerAPITestCocoa.mm:
(WebKit::WebExtensionController::testAdded):
(WebKit::WebExtensionController::testStarted):
...the two new messages are:

the testAdded message, which is sent when browser.test.addTest is called,
and the testStarted message, which is called when the next test is removed
from the queue and has started executing.

* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITest.idl:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
(WebKit::WebExtensionAPITest::assertTrue):
(WebKit::WebExtensionAPITest::assertDeepEq):
(WebKit::WebExtensionAPITest::assertEquals):
(WebKit::WebExtensionAPITest::assertEq):
(WebKit::WebExtensionAPITest::addTest):
(WebKit::WebExtensionAPITest::startNextTest):
(WebKit::assertEquals): Deleted.
Implement the logic for browser.test.addTest, which just maintains a queue of
test methods to run and resolves/rejects the promise based on whether or not
any assertions are hit during the execution.

* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITest.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPITest, AddTestThatPasses)):
(TestWebKitAPI::TEST(WKWebExtensionAPITest, AddTestThatFails)):
Add new tests to ensure that passing and failing tests resolve and reject,
respectively. Additionally, check that the new delegate methods are called
as expected.

* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager initForExtension:extensionControllerConfiguration:]):
(-[TestWebExtensionManager _webExtensionController:recordTestAssertionResult:withMessage:andSourceURL:lineNumber:]):
(-[TestWebExtensionManager _webExtensionController:recordTestEqualityResult:expectedValue:actualValue:withMessage:andSourceURL:lineNumber:]):
(-[TestWebExtensionManager _webExtensionController:recordTestAddedWithTestName:andSourceURL:lineNumber:]):
(-[TestWebExtensionManager _webExtensionController:recordTestStartedWithTestName:andSourceURL:lineNumber:]):
(-[TestWebExtensionManager _webExtensionController:recordTestFinishedWithResult:message:andSourceURL:lineNumber:]):
Throw away test events if we're currently executing a test that was added
with the new browser.test.addTest API. We don't care what the result of
those assertions are, just that the promise resolves/rejects as expected.

Canonical link: https://commits.webkit.org/290558@main
